pkg/routeagent-driver/handlers/kubeproxy-iptables/endpoint_handler.go

```diff
@@ -0,0 +1,169 @@
+package kp_iptables
+
+import (
+	"fmt"
+	"net"
+
+	"k8s.io/klog"
+
+	submV1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/util"
+)
+
+func (kp *SyncHandler) LocalEndpointCreated(endpoint *submV1.Endpoint) error {
+	kp.syncHandlerMutex.Lock()
+	defer kp.syncHandlerMutex.Unlock()
+	kp.localCableDriver = endpoint.Spec.Backend
+
+	// We are on nonGateway node
+	if endpoint.Spec.Hostname != kp.hostname {
+		kp.isGatewayNode = false
+		localClusterGwNodeIP := net.ParseIP(endpoint.Spec.PrivateIP)
+		remoteVtepIP, err := kp.getVxlanVtepIPAddress(localClusterGwNodeIP.String())
+		if err != nil {
+			return fmt.Errorf("failed to derive the remoteVtepIP %v", err)
+		}
+
+		klog.Infof("Creating the vxlan interface %s with gateway node IP %s", VxLANIface, localClusterGwNodeIP)
+		err = kp.createVxLANInterface(VxInterfaceWorker, localClusterGwNodeIP)
+		if err != nil {
+			klog.Fatalf("Unable to create VxLAN interface on non-GatewayNode (%s): %v", endpoint.Spec.Hostname, err)
+		}
+
+		kp.vxlanGwIP = &remoteVtepIP
+		err = kp.reconcileRoutes(remoteVtepIP)
+		if err != nil {
+			return fmt.Errorf("error while reconciling routes %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (kp *SyncHandler) LocalEndpointUpdated(endpoint *submV1.Endpoint) error {
+	return nil
+}
+
+func (kp *SyncHandler) LocalEndpointRemoved(endpoint *submV1.Endpoint) error {
+	kp.syncHandlerMutex.Lock()
+	defer kp.syncHandlerMutex.Unlock()
+	kp.isGatewayNode = false
+
+	if kp.vxlanDevice != nil {
```

Comment:
What if we receive this information out of order?, for example

1) Gateway has moved:

a) new Endpoint is created,
b) old Endpoint is removed.

I think you need to verify if the vxlanDevice you are going to delete was for this endpoint.

Even if probably the order will always be the same, chances are that if the first removal fails for some reason and it's later retried... we could end cleaning up for the wrong endpoint.

Reply:
Done